### PR TITLE
DMIC: Enable capture from secondary FIFO

### PR DESCRIPTION
**Pull request by singalsu**
This patch enables the FIFO B in addition to previously supported
FIFO A. It allows in topology to use in pipelines the DMIC DAI
type with indices 0 and 1. The DAI instances can be operated with
different sample rate and different PCM format. The other
parameters for microphones need to be the same (number of
channels and enabled microphones). If the request via topology
differs for those parameters the latter request overrides the
first request. Typical usage for this added feature is to provide
compact 16 kHz / 16 bit capture version for speech in addition to
normal 48 kHz / 32 bit media quality audio capture.

The DMIC HW actually provides a bit more freedom for FIR/FIFO A/B
usage difference but all of that is not exposed yet via the
driver due to complexity.

The patch modifies two topologies to showcase the feature. The
sof-apl-dmic-2ch/4ch topologies for APL nocodec machine driver
are changed to instantiate in addition PCM7 set for 16 kHz 16 bit
format in addition to earlier PCM6.

Note: The required change to pipe-passthrough-capture.m4 PCM capability
to enable other than 48 kHz rate is not done to not break important
test cases. Another PR addresses the PCM capabilities and after it is
merged allowing 16 kHz PCM for DMIC capture can be done safely.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
@plbossart @lgirdwood @ranj063 We have in IPC two not used parameters: fifo_fs_b and fifo_bits_b. If you prefer I can do another PR set for FW and kernel to remove those (change to reserved) and rename fifo_fs_a -> fifo_fs and fifo_bits_a -> fifo_bits. With the approach here the kernel sees just separate DAIs and does not need to be aware about the dependencies with DMIC FIFOs. At the time of defining the IPC I wasn't sure how this should be done.

**Comment by singalsu** [comment]
@lgirdwood The FIFOs can be set for different sample rates (8 - 96 kHz) as long as the HW internal shared 1st decimation stages can have a common factor and the per FIFO 2nd decimation stages have a low enough integer decimation factor. The FIFO word length can be set always freely to 16 to 32 bits. The embedded FIR coefficients now support decimation by 2, 3, 4, 5, 6, and 8 so quite many sample rate combinations are possible.

I've tested A and B with just one at time running capture plus A and B in various overlapping usage scenarios (assume you meant the latter with A + B). 

The decimation HW consists of 2ch filter blocks called PDM0, PDM1, etc. It would be possible to program the PDM0 when connected only to FIFO A and PDM1 only when connected to FIFO B more independently for e.g. different PDM bus clock but I didn't implement such. If there's requirements it could be done too but would need more work.

**Comment by lgirdwood** [comment]
@singalsu sorry, I mean what combinations of PDM ports can you run with this PR ?

**Comment by singalsu** [comment]
@lgirdwood I've tested with two microphones via PDM0 or PDM1 (2ch) and with four microphones via PDM0 and PDM1 (4ch).

Mono via PDM0 or PDM1 should work too though I didn't test it yet.

**Comment by lgirdwood** [comment]
@singalsu @zrombel can you kick internal CI

File: src/drivers/intel/cavs/dmic.c [hunk 1]
```diff
@@ -1081,38 +1083,26 @@ static int configure_registers(struct dai *dai, struct dmic_configuration *cfg,
...
 static int dmic_set_config(struct dai *dai, struct sof_ipc_dai_config *config)
 {
 	struct dmic_pdata *dmic = dai_get_drvdata(dai);
-	struct sof_ipc_dai_dmic_params *prm;
 	struct matched_modes modes_ab;
 	struct dmic_configuration cfg;
 	struct decim_modes modes_a;
 	struct decim_modes modes_b;
-	int num_pdm_active = config->dmic.num_pdm_active;
+
```
Comment (ranj063) [inline src/drivers/intel/cavs/dmic.c]: @singalsu no need for a line here?

File: src/drivers/intel/cavs/dmic.c [hunk 2]
```diff
@@ -239,23 +243,28 @@ static uint64_t dmic_work(void *data, uint64_t delay)
...
+			case 0:
 				dmic_update_bits(dai, base[i] + FIR_CONTROL_A,
 						 FIR_CONTROL_A_MUTE_BIT, 0);
-
-			if (dmic->fifo_b)
+				break;
+			case 1:
 				dmic_update_bits(dai, base[i] + FIR_CONTROL_B,
 						 FIR_CONTROL_B_MUTE_BIT, 0);
+				break;
```
Comment (ranj063) [inline src/drivers/intel/cavs/dmic.c]: @singalsu no default case here?

File: src/drivers/intel/cavs/dmic.c [hunk 3]
```diff
@@ -239,23 +243,28 @@ static uint64_t dmic_work(void *data, uint64_t delay)
...
 			dmic_write(dai, base[i] + OUT_GAIN_LEFT_A, val);
 			dmic_write(dai, base[i] + OUT_GAIN_RIGHT_A, val);
-		}
-		if (dmic->fifo_b) {
+			break;
+		case 1:
 			val = OUT_GAIN_LEFT_B_GAIN(gval);
 			dmic_write(dai, base[i] + OUT_GAIN_LEFT_B, val);
 			dmic_write(dai, base[i] + OUT_GAIN_RIGHT_B, val);
+			break;
```
Comment (ranj063) [inline src/drivers/intel/cavs/dmic.c]: same here?

File: src/platform/cannonlake/include/platform/interrupt.h [hunk 4]
```diff
@@ -85,7 +85,7 @@
 /* IRQ Level 5 bits */
 #define IRQ_BIT_LVL5_LP_GP_DMA1	15
 #define IRQ_BIT_LVL5_LP_GP_DMA0	16
-#define IRQ_BIT_LVL5_DMIC		6
+#define IRQ_BIT_LVL5_DMIC(x)		(6 + x)
```
Comment (RanderWang) [inline src/platform/cannonlake/include/platform/interrupt.h]: I check the CNL and APL spec, on APL, DMIC interrupt is at pin 6, but on CNL it should be 8.
please conform it.

File: src/platform/icelake/include/platform/interrupt.h [hunk 5]
```diff
@@ -85,7 +85,7 @@
 /* IRQ Level 5 bits */
 #define IRQ_BIT_LVL5_LP_GP_DMA1	15
 #define IRQ_BIT_LVL5_LP_GP_DMA0	16
-#define IRQ_BIT_LVL5_DMIC		6
+#define IRQ_BIT_LVL5_DMIC(x)		(6 + x)
```
Comment (RanderWang) [inline src/platform/icelake/include/platform/interrupt.h]: please also check ICL spec

File: src/platform/apollolake/include/platform/interrupt.h [hunk 6]
```diff
@@ -84,7 +84,7 @@
 /* IRQ Level 5 bits */
 #define IRQ_BIT_LVL5_LP_GP_DMA1(x)	(24 + x)
 #define IRQ_BIT_LVL5_LP_GP_DMA0(x)	(16 + x)
-#define IRQ_BIT_LVL5_DMIC		6
+#define IRQ_BIT_LVL5_DMIC(x)		(6 + x)
```
Comment (RanderWang) [inline src/platform/apollolake/include/platform/interrupt.h]: here, I discuss with keyon. In spec, we only find interrupt pin 6 is for DMIC. pin 7 is for slimbus.
we are confusing about it.

**Comment by zrombel** [comment]
@lgirdwood Done.

File: src/platform/apollolake/include/platform/interrupt.h (hunk 6 above)
Comment (lgirdwood) [reply src/platform/apollolake/include/platform/interrupt.h]: @RanderWang APL and CNL have different IRQ mappings at level 5.

**Comment by ranj063** [comment]
> > @plbossart @lgirdwood @ranj063 We have in IPC two not used parameters: fifo_fs_b and fifo_bits_b. If you prefer I can do another PR set for FW and kernel to remove those (change to reserved) and rename fifo_fs_a -> fifo_fs and fifo_bits_a -> fifo_bits. With the approach here the kernel sees just separate DAIs and does not need to be aware about the dependencies with DMIC FIFOs. At the time of defining the IPC I wasn't sure how this should be done.
> 
> @singalsu do you really want to remove these? If you remove these, how will you track the fs and fifo_bits for the 2 DAI's? Also, if the DMIC params should not be different for the 2 dai's, does it make sense to send just one ipc instead of two?

@singalsu like we discussed today, lets update the IPC structure to remove fifo_b related fields. And also update the global dmic prm to an array to store the config for the 2 FIFO's separately. Thanks!

File: src/drivers/intel/cavs/dmic.c (hunk 2 above)
Comment (singalsu) [reply src/drivers/intel/cavs/dmic.c]: The index has been checked earlier in dmic_set_config() so the check would be redundant. There's quite many similar parts in code so tracing an error in every usage would add code size. Does our coding style require such?

File: src/platform/cannonlake/include/platform/interrupt.h (hunk 4 above)
Comment (singalsu) [reply src/platform/cannonlake/include/platform/interrupt.h]: Yep, good catch!

File: src/platform/icelake/include/platform/interrupt.h (hunk 5 above)
Comment (singalsu) [reply src/platform/icelake/include/platform/interrupt.h]: OK, I just asked help via email.

File: src/platform/apollolake/include/platform/interrupt.h (hunk 6 above)
Comment (singalsu) [reply src/platform/apollolake/include/platform/interrupt.h]: Yep, I have information for APL and CNL. I'll find out also ICL.

Comment (singalsu) [reply src/platform/apollolake/include/platform/interrupt.h]: I'll fix also Sue. ICL is still open, if no info I'll keep it as CNL.

**Comment by singalsu** [comment]
> @singalsu
> You need to increase SOF_TEXT_SIZE by one page for CNL, because of: `` xt-ld: sof section `.text' will not fit in region `sof_text' ``

@jajanusz I'm not able to repeat this. I just tried both gcc and xt-xcc build for SOF.

**Comment by jajanusz** [comment]
> > @singalsu
> > You need to increase SOF_TEXT_SIZE by one page for CNL, because of: `` xt-ld: sof section `.text' will not fit in region `sof_text' ``
> 
> @jajanusz I'm not able to repeat this. I just tried both gcc and xt-xcc build for SOF.

But it passes, so all good

**Comment by singalsu** [comment]
@jajanusz Yep, maybe some merged patch has reduced code size.

**Comment by singalsu** [comment]
> @singalsu like we discussed today, lets update the IPC structure to remove fifo_b related fields. And also update the global dmic prm to an array to store the config for the 2 FIFO's separately. Thanks!

@ranj063 I did first version of FIFO B cleanup from IPC with internal global request storing struct that was much like previous IPC with A and B. It was a simple change.

Then I made another branch for trying an array for storing requests. This approach has the inconvenience with setting up dmic_prm[]  due to the flexible array member in IPC. I allocate double size buffer to dmic_prm[0] and point dmic_prm[1] to buffer half. If I get this version of code to work I'll probably push it to PR. But if it gets too ugly or larger I have the first approach available.

**Comment by singalsu** [comment]
I just pushed update. It was a quite big change over previous PR so hope there's in CI test cases for DMIC and tests for all platform builds. I've tested here single and overlapping cases of 48 and 16 kHz with APL with 4ch and 2ch (via PDM0 and PDM1) that worked.

This will need an IPC cleanup path for Linux as well. I bumped only ABI minor since the change is compatible also without Linux side update.

**Comment by singalsu** [comment]
CI does it's job. I need to fix compile for Sue Creek, stay tuned...

**Comment by singalsu** [comment]
Sue compile issue is now fixed. Hope the builds work OK with this version.

**Comment by lgirdwood** [comment]
@singalsu can you check CI code scan.

**Comment by singalsu** [comment]
@lgirdwood The CI part that failed is a server issue. The server will re-run all failed PRs once it's fixed.

**Comment by singalsu** [comment]
I'm still waiting my access to CI to work. Here's for re-test a fix for what I think caused the check fail.

**Pull request merged**